Minimal TypeScript support option added

Igogrek opened the pull request:
This PR adds TypeScript language option to webpack template.

This is basically the same idea as in: https://github.com/vuejs-templates/webpack/pull/781

The main difference is that I've tried to find minimal possible configuration and code changes.
So in this case there's a metalsmith plugin that renames .js files from /src to .ts if typescript option is enabled.
This way there's no need to maintain both .js and .ts sources.
Checked with both TS enabled and disabled.

nickmessing commented:
@Igogrek, I think we should wait for vuejs/vue#5887 before even considering adding typescript to webpack template, and even then I don't think this will be a valid option since we try to keep this template minimal and even vuex is not here.

Igogrek commented:
Well that's sad. While I certainly agree on keeping template as minimal as possible - TypeScript is one of the most needed (or at least most discussed) features missing on vue-cli (see https://github.com/vuejs/vue-cli/issues/263)

https://github.com/vuejs/vue/pull/5887 shouldn't break anything but I will double check

nickmessing commented:
@Igogrek, I think with all of discussion around TypeScript we can make a dedicated typescript template, @posva, do you think that's a valid option after vuejs/vue#5887 gets merged?

P.S. I really think  vuejs/vue#5887 will be a big change for typescript support since it will make it possible to not use `vue-class-component`. I'm waiting for it for a long time.

LinusBorg commented:
@nickmessing makes sense to wait for #5887.

yalishizhude commented:
Hey~you forget unit tests

Igogrek commented:
@yalishizhude 
Thanks. Will think how to do it best after https://github.com/vuejs/vue/pull/6391 is merged.

Igogrek commented:
Updated the fork and dependencies versions.

@yalishizhude 
I've added unit test TypeScript support. Note that it will need to be changed again as https://github.com/vuejs-templates/webpack/pull/824 likely will be merged soon.

Waiting for 2.5 for now, to be able to update TypeScript usages properly (definitions are merged already).

solarhell commented:
@Igogrek Vue [2.5.0](https://github.com/vuejs/vue/releases/tag/v2.5.0) already released 🍺

Igogrek commented:
@solarhell I've updated PR again to use Vue.extend() now.

Couldn't find any docs for the new intended TypeScript definition usage, but this approach seems best and autocompletion works great.

Tested different combinations with different ESLint settings and everything seem to work now.

championswimmer commented on template/tsconfig.json:
```diff
@@ -0,0 +1,16 @@
+  "compilerOptions": {
+    "module": "es2015",
+    "moduleResolution": "node",
+    "allowSyntheticDefaultImports": true,
```
not needed anymore I presume

igogrek replied:
Yes, removed. Thanks.

asfer80 commented:
For anyone wanting to give this PR a try you can use:

`vue init vuejs-templates/webpack#pull/797/head my-project`

skray commented:
@Igogrek did you have plans to convert the included unit test as well? I would find it strange to generate a TS project and to have some of the files still be Javascript.

I bring this up because if you do convert it over, at a minimum you'll need to include the `@types/mocha` and `@types/chai` dependencies, and add `import {expect} from 'chai'` to the `Hello.spec.ts` file.

Igogrek commented:
@skray Thanks. Totally forgot to add chai - fixed now.

njleonzhang commented:
@Igogrek, I get warning: `can not find module '@/components/HelloWorld'`, can you help?

```
import HelloWorld from '@/components/HelloWorld'
```

Igogrek commented:
@njleonzhang 
Just checked - seem to be fine for me.
I assume you have a warning in IDE, not the project build?

njleonzhang commented:
@Igogrek Yeah in `vscode`. But the build is OK.

Igogrek commented:
@njleonzhang You're right. Guess I'll need to put back vue-shims for better IDE support.

For now put `vue-shims.d.ts` in `src` folder with:
```
declare module "*.vue" {
    import Vue from "vue";
    export default Vue;
}
```
And change imports to have a `.vue` extension:
```
import HelloWorld from '@/components/HelloWorld.vue'
...
import App from './App.vue'
```

Since I don't really like leaving this extensions only for TypeScript - I'll try to find other solution which works both for CLI build and IDE too.

njleonzhang commented:
@Igogrek Tnx, I also found this solution. As you mentioned. It's not good enough. Hope you can find perfect one. 
BTW, I have a question on typescript and mixin. would you please help to check? 
https://forum.vuejs.org/t/typescript-to-restrict-vue-component-instance-must-have-special-properties-or-methods/21709

Igogrek commented:
@njleonzhang No, sorry can't help with Mixin.

For the IDE support - https://github.com/vuejs/vetur/issues/213 indicates that  `.vue` extension is the only solution for now. Reverted shims back.

RokkerRuslan commented:
So, when the merge?

Right2Drive commented:
I know I tried getting this to work with Jest earlier in the month and failed, due to the limited support offered by Vue and Jest for Typescript. Something such as [jestpack](https://github.com/60frames/jestpack) may offer a solution to this, although I'm not sure if this is as reliable as [ts-jest](https://github.com/kulshekhar/ts-jest), and definitely less reliable than jest itself. I could see this being a major hurdle in integrating this with #824

Igogrek commented:
@Right2Drive Thanks for the info.
jestpack looks quite complex to setup and very outdated. ts-jest turned out to be just the right tool.
I've added basic Jest support for now, coverage should be tweaked though (will do later).

Overall this PR is getting harder and harder for me alone to maintain.

@LinusBorg is there any chance this getting merged?

lsim commented:
Not sure if this sort of thing is appropriate here, but I have spent a couple of days converting an angular2 project into vue+ts using this template and it's been a very good experience. This a seriously powerful stack - and simple enough in terms of configuration that you can adapt it to your needs - I like the balance it strikes.

Just wanted to say thanks for the hard work - I hope you get it merged soon!

sthomp commented:
Have you considered adding [tslint](https://github.com/palantir/tslint) to the template?

Igogrek commented:
@sthomp 
I did, but at the time of me checking it TSLint had some issues with .vue files.
And it seems that issue is still open https://github.com/palantir/tslint/issues/2099

ESLint worked ok for me, additionally TS plugin could be installed. But ideally for TypeScript - TSLint should be used.
I will check again soon.

Igogrek commented:
I've updated the fork again. 
Had some more issues with Jest+TypeScript but resolved them now.
I've also fixed unit tests so it builds successfully now.

@LinusBorg since there are unit tests now, can this be merged or not?

RokkerRuslan commented:
It is the magic PR, maintainer (@LinusBorg ) don`t see that.

LinusBorg commented:
Hey everyone. 

We appreciate the hard work that was put into this  - and I will consider a merge, just have to take it for a spin myself before to get a feel how much complexity it adds. 

We originally planned a separate template, but that won't happen anymore before we do vue-cli 3.0 - which itself is still in the concept phase. 

So I understand that an official integration would be welcome by many people, and I don't want to let this work go to waste. 

The only thing that's might end up in a "no" vote would be if this makes the maintenance too complicated - I'll see about that. 

Unfortunately I I'll be travelling this weekend to won't have a chance to check this out with the dilligence that's reuqired.

ThePlastic commented:
@Igogrek take into consideration the idea of adding support to sass and scss, in my case it was enough to add these 2 lines to package.json's dev dependencies:
`"sass-loader": "^ 6.0.6"`
`"node-sass": "^ 4.7.2"`

and the I can simply use: `<style lang="scss">` and it compiles without any problem.

LinusBorg commented:
@ThePlastic then to be "fair", packages for all supported preprocessors (sass/scss, less, stylus) should be added - which imho bloats dependencies unnecessarily, but that'S debatable.

ThePlastic commented:
@LinusBorg i can see your point

Igogrek commented:
I've updated the PR and checked with both TS enabled and disabled - everything seem to be fine.

@LinusBorg did you have time to look at the PR locally?

LinusBorg commented:
I have taken a first look, and it's looking good for the most part. Have some small points/questions but probably won'T be able to write them down till Thursday evening.

lyonsbp commented:
First of all thanks for the good work here! So I just installed this with `vue init vuejs-templates/webpack#pull/797/head project-name` and have a Typescript error in the HelloWorld.vue file.
`Cannot find module './components/HelloWorld' (10,24)`. This is in VS Code if that matters. The code still compiles fine when I run npm dev, but I'd like to get rid of the error if possible.

lsim commented:
Yeah, I've seen this too. Vscode wants you to include the  .vue extension when importing components.

lyonsbp commented:
Well I have a different project that is working with .vue files which is why im confused. This is the working project, this is the working project: https://github.com/lyonsbp/vue-typescript-spa 
Which is just a remake of the official Microsoft Typescript Vue starter, which is [here](https://github.com/Microsoft/TypeScript-Vue-Starter#typescript-vue-starter)

lyonsbp commented:
@Igogrek Could you please add the .vue extension to the App.vue file? I don't know how to make a PR on a PR, new to git sorry.

Igogrek commented:
@lyonsbp I've added the extension - should not be a problem anymore (was only the case if router isn't installed).

yyx990803 commented:
Thanks for the great work @Igogrek - this is actually quite helpful as I am designing the architecture for the next iteration of `vue-cli`. The core problem of this template right now, is that there are multiple options, and each option has its changes scattered all over the template (e.g. the number of files touched for adding minimal TS support). Hopefully we'll be address that in the next version of the CLI so all TS-related changes can be centralized and maintained in a single place.

In the meanwhile, I think the complexity introduced by this PR is manageable for the period of time before the new CLI is available.

lzl124631x commented:
can we achieve simply adding `<script lang="typescript">` to enable typescript in `.vue`?

LinusBorg commented:
More or less, but not really. 

* the main entry point has to be .ts
* tsconfig.json should be provided 
* typescript dependencies are not installed

Just check out the diff to see what UP did. 

BTW once #1218 is merged I will merge this next. 

I just wanted to have a flexible testing setup first to test the template with and without typescript in CI.

prograhammer commented:
Hi guys,

This [PR](https://github.com/Realytics/fork-ts-checker-webpack-plugin/pull/77) for **fork-ts-checker-weback-plugin** will TypeCheck and TSLint the typescript blocks in Vue SFCs. We're hoping to merge it in soon but you can test it out with the built repo: `npm install git://github.com/prograhammer/fork-ts-checker-webpack-plugin.git#vue --save-dev`.

Also, use the shorter `lang="ts"` in the script section (I'll update that later). Check out the first post of that PR for complete information.

Toilal commented:
`fork-ts-checker-weback-plugin` has been released as 0.3.0 with support for vue SFC files.

Toilal commented:
When TypeScript is chosen, `.vue` files should not be processed by eslint. Another option is to include https://github.com/nzakas/eslint-plugin-typescript : This should allow to run "eslint --fix" after generating the project

Igogrek commented:
@Toilal 
That could work, but I think that TSLint via `fork-ts-checker-weback-plugin` would be much better choice since it can do TSLint too.

I would've added this plugin already but most likely there will be more issues with specific cases (like I already had with Jest) and I would rather leave this as is for now.
Or it would be nice if someone could test this out thoroughly.

ffxsam commented:
@Igogrek For what it's worth, feel free to borrow any of my work from this PR: https://github.com/vuejs-templates/webpack/pull/1238

It adds TSLint support to the template, with the right configuration to have it work out of the box with VSCode. It was a bit tricky, since TSLint's `no-unused-variables` seems to fight with the tsconfig `noUnusedLocals`.

Toilal commented:
fork-ts-checker-webpack-plugin is really better, but it involves more changes. You may have a look to http://github.com/Toilal/vue-webpack-template as it's already implemented among other options. (it's a fork and I try merge as often as possible to minimize changes from official)

igogrek commented:
@LinusBorg 
I'm not completely sure how to fix full and full-karma-airbnb builds as they seem to miss some scripts, namely `lint` `unit` and `e2e` (I assume not in package.json).
Just updated from `develop` and it doesn't look like it's caused by my code..

SYip commented on template/src/main.js:
```diff
@@ -3,7 +3,7 @@
 // (runtime-only or standalone) has been set in webpack.base.conf with an alias.
```
Should change the file name to main.ts?

igogrek replied:
It is changed via metalsmith - that was the basic idea of the PR. 
See my initial comment and `meta.js`

ThePlastic commented:
I wasn't able to try this pull anymore... I got a lot of error and it doesn't ask for any option (i mean, typescript, unit test, linter,  e2e...)

igogrek commented:
@ThePlastic 
Pretty sure this is not related to the PR, but with the local environment - most likely global Vue CLI version was updated (3.0?).

ThePlastic commented:
@Igogrek I'm sure it is.
I've also updated my vue CLI but got the same errors. So this pull is no longer usable until merge?

igogrek commented:
@ThePlastic
If you downgrade your CLI back to the version 2 it should work.

mikob commented:
Thanks for your work on this! Unfortunately, I wasn't able to get it to work either, I've tried vue-cli@2.8.1, 2.9.3 and the current 3 version. Errors look like this: 
```
Error when evaluating filter condition: isNotTest
Error when evaluating filter condition: isNotTest
...
```

igogrek commented:
@mikob
This will not work with version 3 - try using default TS plugin for it.

I will try to update from fresh `develop` soon to get unit tests fixed but for now I don't have time.

Patskimoto commented:
Just wondering what's the current status on this being merged into master? I am beginning a new project that will eventually be significant in size, I love all the features this package includes and would really like to use TypeScript within the project.

ffxsam commented:
@Patskimoto Why wouldn't you just use the latest vue-cli which has built-in TypeScript support?

Patskimoto commented:
@ffxsam  I have since done just that - thank you.
